crates/core/src/client/helpers.rs

```diff
@@ -64,36 +69,38 @@ impl TryFrom<Vec<FieldElement>> for Calls {
     }
 }
 
-/// Returns the decoded return value of the `eth_call` entrypoint of Kakarot
+/// Returns the decoded return value of the `eth_call` and `eth_send_transaction` entrypoint of
+/// Kakarot
 pub fn decode_eth_call_return<T: std::error::Error>(
     call_result: &[FieldElement],
 ) -> Result<Vec<FeltOrFeltArray>, EthApiError<T>> {
-    // Parse and decode Kakarot's call return data (temporary solution and not scalable - will
+    // Parse and decode Kakarot's return data (temporary solution and not scalable - will
     // fail is Kakarot API changes)
-    // Declare Vec of Result
-    let mut segmented_result: Vec<FeltOrFeltArray> = Vec::new();
-    let mut tmp_array_len: FieldElement = *call_result.first().ok_or_else(|| {
-        EthApiError::OtherError(anyhow::anyhow!("Cannot parse and decode return arguments of Kakarot call",))
-    })?;
-    let mut tmp_counter = 1_usize;
-    segmented_result.push(FeltOrFeltArray::FeltArray(Vec::new()));
-    // Parse first array: stack_accesses
-    while tmp_array_len != FieldElement::ZERO {
-        let element = call_result.get(tmp_counter).ok_or_else(|| {
-            EthApiError::OtherError(anyhow::anyhow!(
-                "Cannot parse and decode return arguments of Kakarot call: return data array",
-            ))
-        })?;
-        match segmented_result.last_mut() {
+
+    let mut return_data: Vec<FeltOrFeltArray> = vec![FeltOrFeltArray::FeltArray(vec![])];
+
+    let return_data_len =
+        *call_result.first().ok_or_else(|| DataDecodingError::InvalidReturnArrayLength("call result length".into()))?;
+
+    let mut return_data_len: u64 =
+        return_data_len.try_into().map_err(|e: ValueOutOfRangeError| ConversionError::Other(e.to_string()))?;
+    let mut counter = 1_usize;
+
+    // Parse call result array
+    while return_data_len != 0 {
+        let element = call_result
+            .get(counter)
+            .ok_or_else(|| DataDecodingError::InvalidReturnArrayLength("call result".into()))?;
+        match return_data.last_mut() {
             Some(FeltOrFeltArray::FeltArray(felt_array)) => felt_array.push(*element),
             Some(FeltOrFeltArray::Felt(_felt)) => (),
             _ => (),
         }
-        tmp_counter += 1;
-        tmp_array_len -= FieldElement::from(1_u64);
+        counter += 1;
+        return_data_len -= 0;
```

Comment:
if we're reducing the length by 0, wont this be an infinite loop?

Reply:
yeah my bad, I fix this in another [PR](https://github.com/kkrt-labs/kakarot-rpc/pull/244/files#diff-9ad132b6b358cc76d7905cfa6d53d37cb3b8b28c2fcf0dc5f5bbfcf7dd0f45ab) and add a test so this won't happen again.